src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1234,10 +1236,121 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
 		}
 	}
 
+	/*
+	 * We could stil behave as if the target list is on partition column if
+	 * all range table entries are reference tables and there is no column
+	 * references to any outer queries.
+	 */
+	if (!targetListOnPartitionColumn)
+	{
+		if (RangeTableListContainsOnlyReferenceTables(query->rtable) &&
+			!TargetListContainsNotSupportedEntries(targetEntryList))
+		{
+			targetListOnPartitionColumn = true;
+		}
+	}
+
 	return targetListOnPartitionColumn;
 }
 
 
+/*
+ * TargetListContainsNotSupportedEntries returns true if non of the
+ * elements in the target entry list belong to an outer query (for
+ * example the query is a sublink and references to another query
+ * in the from list).
+ *
+ * The function also returns true if any of the  target entries is not
+ * a column itself. This might be too restrictive, but, given that we're
+ * handling a very specific type of queries, that seems acceptable for now.
+ */
+static bool
+TargetListContainsNotSupportedEntries(List *targetEntryList)
+{
+	ListCell *targetEntryCell = NULL;
+
+	foreach(targetEntryCell, targetEntryList)
+	{
+		TargetEntry *targetEntry = lfirst(targetEntryCell);
+		Var *candidateColumn = NULL;
+
+		Expr *strippedColumnExpression = (Expr *) strip_implicit_coercions(
+			(Node *) targetEntry->expr);
+
+		if (IsA(strippedColumnExpression, Var))
+		{
+			candidateColumn = (Var *) strippedColumnExpression;
+		}
+		else if (IsA(strippedColumnExpression, FieldSelect))
+		{
+			FieldSelect *compositeField = (FieldSelect *) strippedColumnExpression;
+			Expr *fieldExpression = compositeField->arg;
+
+			if (IsA(fieldExpression, Var))
+			{
+				candidateColumn = (Var *) fieldExpression;
+			}
+		}
+
+		/* we don't support target entries that are not columns */
+		if (candidateColumn == NULL)
+		{
+			return true;
+		}
+
+		if (candidateColumn->varlevelsup > 0)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * RangeTableListContainsOnlyReferenceTables returns true if all relation range
```

Comment:
possibly rephrase **if all relation range  tables entries are reference tables.** to **if all range table entries are reference tables.**

Reply:
Oh, sure we should do that